engine/src/cubos/engine/input/bindings.cpp

```diff
@@ -0,0 +1,202 @@
+#include <cubos/core/log.hpp>
+
+#include <cubos/engine/input/bindings.hpp>
+
+using cubos::core::data::Deserializer;
+using cubos::core::data::Serializer;
+using cubos::core::io::Key;
+using cubos::core::io::keyToString;
+using cubos::core::io::Modifiers;
+using cubos::core::io::modifiersToString;
+using cubos::core::io::stringToKey;
+using cubos::core::io::stringToModifiers;
+using namespace cubos::engine;
+
+const Key& KeyWithModifiers::key() const
+{
+    return mKey;
+}
+
+const Modifiers& KeyWithModifiers::modifiers() const
+{
+    return mModifiers;
+}
+
+Key& KeyWithModifiers::key()
+{
+    return mKey;
+}
+
+Modifiers& KeyWithModifiers::modifiers()
+{
+    return mModifiers;
+}
+
+const std::vector<KeyWithModifiers>& InputAxis::positive() const
+{
+    return mPositive;
+}
+
+const std::vector<KeyWithModifiers>& InputAxis::negative() const
+{
+    return mNegative;
+}
+
+std::vector<KeyWithModifiers>& InputAxis::positive()
+{
+    return mPositive;
+}
+
+std::vector<KeyWithModifiers>& InputAxis::negative()
+{
+    return mNegative;
+}
+
+float InputAxis::value() const
+{
+    return mValue;
+}
+
+void InputAxis::value(float value)
+{
+    if (std::abs(value) > 1.0f)
+    {
+        CUBOS_WARN("Axis value out of range: {}", value);
+    }
+    mValue = std::clamp(value, -1.0f, 1.0f);
+}
+
+const std::vector<KeyWithModifiers>& InputAction::keys() const
+{
+    return mKeys;
+}
+
+std::vector<KeyWithModifiers>& InputAction::keys()
+{
+    return mKeys;
+}
+
+bool InputAction::pressed() const
+{
+    return mPressed;
+}
+
+void InputAction::pressed(bool pressed)
+{
+    mPressed = pressed;
+}
+
+const std::unordered_map<std::string, InputAction>& InputBindings::actions() const
+{
+    return mActions;
+}
+
+const std::unordered_map<std::string, InputAxis>& InputBindings::axes() const
+{
+    return mAxes;
+}
+
+std::unordered_map<std::string, InputAction>& InputBindings::actions()
+{
+    return mActions;
+}
+
+std::unordered_map<std::string, InputAxis>& InputBindings::axes()
+{
+    return mAxes;
+}
+
+template <>
+void cubos::core::data::serialize<InputBindings>(Serializer& ser, const InputBindings& obj, const char* name)
+{
+    ser.beginObject(name);
+    ser.beginDictionary(obj.actions().size(), "actions");
+    for (const auto& [actionName, action] : obj.actions())
+    {
+        ser.write(actionName, nullptr);
+        ser.write(action.keys(), nullptr);
+    }
+    ser.endDictionary();
+    ser.beginDictionary(obj.axes().size(), "axes");
+    for (const auto& [axisName, axis] : obj.axes())
+    {
+        ser.write(axisName, nullptr);
+        ser.beginObject(nullptr);
+        ser.write(axis.positive(), "pos");
+        ser.write(axis.negative(), "neg");
+        ser.endObject();
+    }
+    ser.endDictionary();
+    ser.endObject();
+}
+
+template <>
+void cubos::core::data::deserialize<InputBindings>(Deserializer& des, InputBindings& obj)
+{
+    des.beginObject();
+
+    size_t actions_sz = des.beginDictionary();
```

Comment:
Always prefix `size_t` with `std::` - the standard does not require `size_t` to be defined in the global scope, we had problems with this in the past.